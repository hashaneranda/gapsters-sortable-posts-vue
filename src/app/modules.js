import posts from "../features/posts";

export default {
  posts,
};
