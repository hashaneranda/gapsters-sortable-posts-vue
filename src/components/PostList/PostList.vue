<template>
  <div data-testid="post-list" class="container mx-auto" :class="className">
    <h1 class="text-4xl leading-loose text-white">Sortable Post List</h1>
    <Post
      v-for="(post, index) in posts"
      :key="post.id"
      :id="post.id"
      :isStart="index === 0"
      :isEnd="index === posts.length - 1"
      :buttonUp="() => handleButtonClick(index, 'up', post.id)"
      :buttonDown="() => handleButtonClick(index, 'down', post.id)"
    />
  </div>
</template>

<script>
import Post from "../Post/Post";

//utils
import { swapArrayElements } from "./utils/ArrayHelper";

//vuex
import $store from "@/app/store";

export default {
  name: "PostList",
  props: {
    className: String,
  },
  components: {
    Post,
  },
  computed: {
    posts: () => {
      return $store.state.posts.postList.data;
    },
  },
  methods: {
    handleButtonClick(id, action, postId) {
      if (action === "up") {
        const newList = swapArrayElements([...this.posts], id - 1, id);
        const action = `Moved post ${postId} from index ${id} to index ${id -
          1}`;

        $store.commit("posts/ADD_TO_ACTION_STACK", {
          action,
          order: this.posts,
        });
        $store.commit("posts/MOVE_UP_POST", newList);
      } else if (action === "down") {
        const newList = swapArrayElements([...this.posts], id, id + 1);
        const action = `Moved post ${postId} from index ${id} to index ${id +
          1}`;

        $store.commit("posts/ADD_TO_ACTION_STACK", {
          action,
          order: this.posts,
        });
        $store.commit("posts/MOVE_DOWN_POST", newList);
      }
    },
  },
  created() {
    $store.dispatch("posts/FETCH_POSTS");
  },
};
</script>
