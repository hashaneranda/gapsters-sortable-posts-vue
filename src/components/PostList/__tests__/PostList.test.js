import Vuex from "vuex";
import { render, cleanup } from "@testing-library/vue";
import { createLocalVue, mount } from "@vue/test-utils";

import PostList from "../PostList";

import modules from "@/app/modules.js";

const localVue = createLocalVue();
localVue.use(Vuex);

const store = new Vuex.Store({
  modules,
});

afterEach(cleanup);

describe("PostList Componenet", () => {
  it("should take a snapshot", () => {
    const wrapper = mount(PostList, {
      store,
      localVue,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders componenet corretly without passing argumenets", () => {
    const { getByTestId } = render(PostList, {
      store,
      localVue,
    });
    expect(getByTestId("post-list")).toBeInTheDocument();
  });
});
