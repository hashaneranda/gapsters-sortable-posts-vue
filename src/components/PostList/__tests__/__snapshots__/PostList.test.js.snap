// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList Componenet should take a snapshot 1`] = `
<div
  class="container mx-auto"
  data-testid="post-list"
>
  <h1
    class="text-4xl leading-loose text-white"
  >
    Sortable Post List
  </h1>
   
</div>
`;
